pkg/cmd/stack/deploy.go

```diff
@@ -15,115 +15,170 @@ package stack
-	if err := translate(ctx, s, forceBuild, noCache); err != nil {
+	c, _, err := client.GetLocal()
+	if err != nil {
 		return err
 	}
 
-	dynamicStackFilename, err := saveStackFile(s)
-	if err != nil {
+	cfg := translateConfigMap(s)
+	output := fmt.Sprintf("Deploying stack '%s'...", s.Name)
+	cfg.Data[statusField] = progressingStatus
+	cfg.Data[outputField] = base64.StdEncoding.EncodeToString([]byte(output))
+	if err := configmaps.Deploy(ctx, cfg, s.Namespace, c); err != nil {
 		return err
 	}
-	defer os.Remove(dynamicStackFilename)
 
-	valueOpts := &values.Options{}
-	valueOpts.ValueFiles = []string{dynamicStackFilename}
-	vals, err := valueOpts.MergeValues(nil)
+	err = deploy(ctx, s, forceBuild, wait, noCache, c)
 	if err != nil {
-		return fmt.Errorf("error initializing stack values: %s", err)
+		output = fmt.Sprintf("%s\nStack '%s' deployment failed: %s", output, s.Name, err.Error())
+		cfg.Data[statusField] = errorStatus
+		cfg.Data[outputField] = base64.StdEncoding.EncodeToString([]byte(output))
+	} else {
+		output = fmt.Sprintf("%s\nStack '%s' successfully deployed", output, s.Name)
+		cfg.Data[statusField] = deployedStatus
+		cfg.Data[outputField] = base64.StdEncoding.EncodeToString([]byte(output))
 	}
 
-	var re *repo.Entry
-	rf, err := repo.LoadFile(settings.RepositoryConfig)
-	if !isNotExist(err) {
-		for _, r := range rf.Repositories {
-			if r.Name != stackHelmRepoName {
-				continue
-			}
-			re = r
-			break
-		}
-	}
-	if re == nil {
-		if err := helm.AddRepo(settings, stackHelmRepoName, stackHelmRepoURL, stackHelmChartName, stackHelmChartVersion); err != nil {
-			return err
-		}
-		log.Information("'%s' has been added to your helm repositories.", stackHelmRepoName)
-	} else {
-		err := helm.UpdateRepo(re, settings, stackHelmChartName)
-		if err != nil {
-			return err
-		}
+	if err := configmaps.Deploy(ctx, cfg, s.Namespace, c); err != nil {
+		return err
 	}
 
+	return err
+}
+
+func deploy(ctx context.Context, s *model.Stack, forceBuild, wait, noCache bool, c *kubernetes.Clientset) error {
 	spinner := utils.NewSpinner(fmt.Sprintf("Deploying stack '%s'...", s.Name))
 	spinner.Start()
 	defer spinner.Stop()
 
-	actionConfig := new(action.Configuration)
-	if err := actionConfig.Init(settings.RESTClientGetter(), s.Namespace, helmDriver, func(format string, v ...interface{}) {
-		message := strings.TrimSuffix(fmt.Sprintf(format, v...), "\n")
-		spinner.Update(fmt.Sprintf("%s...", message))
-	}); err != nil {
-		return fmt.Errorf("error initializing stack client: %s", err)
+	if err := translate(ctx, s, forceBuild, noCache); err != nil {
+		return err
 	}
 
-	exists, err := helm.ReleaseExist(action.NewList(actionConfig), s.Name)
-	if err != nil {
-		return fmt.Errorf("error listing stacks: %s", err)
+	for name := range s.Services {
+		if len(s.Services[name].Volumes) == 0 {
+			if err := deployDeployment(ctx, name, s, c); err != nil {
+				return err
+			}
+		} else {
+			if err := deployStatefulSet(ctx, name, s, c); err != nil {
```

Comment:
I don't see value in using stateful sets, and I think they limit how we can do updates, as well as our usage of volumes. what if we just support deployments, and we create the PVC?

Reply:
Yes, but it is the recommended way to use pods with volumes. They also make it simple to handle volumes in multi-zone clusters. We can add logic to regenerate the statefulset when needed